src/drivers/infiniband/hermon.c

```diff
@@ -137,13 +137,13 @@ static int hermon_cmd_wait ( struct hermon *hermon,
 			     struct hermonprm_hca_command_register *hcr ) {
 	unsigned int wait;
 
-	for ( wait = HERMON_HCR_MAX_WAIT_MS ; wait ; wait-- ) {
+	for ( wait = 100 * HERMON_HCR_MAX_WAIT_MS ; wait ; wait-- ) {
 		hcr->u.dwords[6] =
 			readl ( hermon->config + HERMON_HCR_REG ( 6 ) );
 		if ( ( MLX_GET ( hcr, go ) == 0 ) &&
 		     ( MLX_GET ( hcr, t ) == hermon->toggle ) )
 			return 0;
-		mdelay ( 1 );
+		udelay ( 10 );
```

Comment:
Does this make a noticeable difference to anything?  On my test system (CX2 EN) I see a total of 44 commands issued to get to the point that the card is sending and receiving packets, which would give you a maximum saving of 44ms.

Reply:
Do you use a dual-port card? I see a lot more activity. I'll try to add some debug stmts to figure out if there's something to it

Reply:
Yeah, just checked. I very quickly got up to tens of thousands of commands.

Whatever logic is polling the link status is... determined.

Reply:
I added this code to `hermon_cmd()`:

```c
	static int num = 0;
	num++;
	TR("Command[%5d]: %08x", num++, opcode);
```

This is the example output after a couple of seconds:

```
[drivers/infiniband/hermon.c/hermon_cmd:211] Command[11755]: 00000024
```

The explosive amount of commands all seem to be opcode 0x24 - `HERMON_HCR_MAD_IFC` ?

This might indicate a deeper problem..

Reply:
`HERMON_HCR_MAD_IFC` is used in `hermon_mad()`

This might trigger it?

```
	/* No event is generated upon reaching INIT, so we must poll
	 * separately for link state changes while we remain DOWN.
	 */
	if ( ib_is_open ( ibdev ) &&
	     ( ibdev->port_state == IB_PORT_STATE_DOWN ) ) {
		ib_smc_update ( ibdev, hermon_mad );
	}
```

That's from `hermon_poll_eq()`.. so on every event queue poll, we ask the firmware to send a mgmt datagram. That doesn't really seem right, does it?

Reply:
This thing about no event generated for reaching "INIT", does that refer to full online infiniband, or physical link-up?

I have verified that (at least CX3) happily generates link state change events when physical link goes up.

I haven't touched this part of the code, btw, so it seems like a "pre-existing condition", as it were. I will concede that it's looking like the command wait logic does not need to be more agressive, if we would stop posting so many commands.

Reply:
I'm doing some more testing to figure out what's up with this

Reply:
I'm pretty sure this is going to be Infiniband-only, which explains why I'm not seeing it when testing on a CX2 EN.

The "no event generated" was definitely a real problem on at least some hardware supported by `hermon.c`: commit https://github.com/ipxe/ipxe/commit/14a76b592 and my handwritten notes from 2010-09-14 confirm the existence of the problem.

So, we **do** need to perform this `ib_smc_update()` polling while DOWN since at least some versions of the supported hardware will required it.  It's not actually sending anything over the wire: it's just sending something that looks like a management datagram to the firmware since that's the method by which software and firmware communicate.

I don't really want to start guessing as to which versions of the hardware (and very possibly firmware) require this workaround, so we're going to have to keep it for all supported cards.

I'll merge this commit with a message updated to explain the situation in which it's needed.

Reply:
Agreed - guessing is never fun.

I'm going to test if CX3 requires this workaround. If not, we could disable it for CX3 cards? I'll do that after the merge then.

Reply:
We could really consider throttling the number of smc updates, in any case. There's probably zero advantage of hammering the firmware with several thousand requests per second. If we did, say, 5-10 checks per second, we wouldn't reasonably need the more agressive cmd wait logic.

Reply:
Reworded as https://github.com/ipxe/ipxe/commit/6cb12ee2b

Reply:
I'm happy to leave it as-is (now that https://github.com/ipxe/ipxe/commit/6cb12ee2b is merged) on the basis of keeping it simple.  In the common case, the link will stay in DOWN for only a very short length of time, unless the user is trying to use a port for which the cable is physically unplugged.